packages/crypto/src/crypto.ts

```diff
@@ -23,3 +23,45 @@ export const hashLeftRight = (input1: any, input2: any) =>
 export const hashOne = (input: any) => hash1([input])
 
 export { SNARK_FIELD_SIZE, genRandomSalt, stringifyBigInts, unstringifyBigInts }
+
+export const genEpochKey = (
+    identityNullifier: bigint,
+    attesterId: bigint | string,
+    epoch: bigint | number,
+    nonce: bigint | number,
+    maxEpochKey: bigint | number
+): bigint => {
+    const epochKey = hash4([
+        identityNullifier as any,
+        BigInt(attesterId),
+        epoch,
+        BigInt(nonce),
+    ]).valueOf()
+    // Adjust epoch key size according to epoch tree depth
+    const epochKeyModed = epochKey % BigInt(maxEpochKey)
+    return epochKeyModed
+}
+
+export const genEpochNullifier = (
+    identityNullifier: bigint,
+    attesterId: bigint | string,
+    epoch: number | bigint
+): bigint => {
+    return hash3([BigInt(attesterId), BigInt(epoch), identityNullifier as any])
+}
+
+export const genStateTreeLeaf = (
+    idNullifier: bigint,
+    attesterId: bigint | string,
+    epoch: bigint | number,
+    posRep: bigint | number,
+    negRep: bigint | number
+): bigint => {
+    return hash5([
+        idNullifier,
+        BigInt(attesterId),
+        BigInt(epoch),
+        BigInt(posRep),
+        BigInt(negRep),
+    ])
+}
```

Comment:
these two functions are also more protocol related so that it should be in `@unirep/core`